### ISSUE TITLE
Add option for daemonizing crystal playground

### ISSUE THREAD

**Issue by Sija** [crystal-lang/crystal#3162]
Something along `crystal play --daemonize` would be useful for "run & forget" scenarios when you want to have the playground running all the time in the bg.

**Comment by sumproxy** [comment]
Why not just `crystal play&` ?

**Comment by Sija** [comment]
@sumproxy Cause it will exit along with the current terminal session.

**Comment by jhass** [comment]
Well you can just 
```
crystal play &
nohup crystal play
systemd-run --user crystal play

**Comment by Sija** [comment]
@jhass That's true, but it feels to me kinda "hackish" - for instance, stopping the server would require killing it manually. IMO would be nice to have baked-in, OS-independent solution, without resorting to `systemd`, `supervisor` or similar tools.

**Comment by jhass** [comment]
Actually using a supervisor gives you much saner tools to kill it. Self-daemonizaition requires to kill it just the same way as `nohup` does. With `&` you can just `fg` and Ctrl-C it.

**Comment by Sija** [comment]
Sure, but would it hurt to make a thin wrapper around it?
- `crystal play start`: nohup crystal play
- `crystal play stop`: kill nohup-ed *pid*

It feels IMO more natural and doesn't require a user to know the underlying, albeit simple, impl.

**Comment by jhass** [comment]
I'm on the fence for user facing daemons.

For server daemons I fully agree with http://www.mikeperham.com/2014/09/22/dont-daemonize-your-daemons/

With systemd user sessions becoming a fairly common thing these days, I tend to think the same arguments start to apply here too.

**Comment by Sija** [comment]
For server daemons, I'd agree that using hand-crafted solutions is usually sub-optimal but for development, it's a frequently chosen path—see for instance Rails' app preloader [Spring](https://github.com/rails/spring).

**Comment by ysbaddaden** [comment]
Actually I believe in the opposite: a simple `-d` for user daemon's is easier than relying on a long commands that depends on your init system.

- a one letter arg that works on every OS/init system;
- the process crashes? manually start it again;
- your daemon process isn't killed when you log out? but who logs out from their computer? we shutdown or put to sleep/hibernation.

**Comment by RX14** [comment]
Adding `& disown` to the end of the command isn't much longer. Shell background jobs were built for this.

**Comment by Sija** [comment]
@RX14 `nohup` would work better here since it separates the process from the terminal, allowing to safely close it and keep the process running in the background.

Even so, I'd still favor using `-d` option for run-and-forget scenario, instead of using `nohup` and killing the process via things like `ps -ef | awk '$8=="crystal" && $9=="play" {print $2}' | xargs kill`

**Comment by RX14** [comment]
`killall crystal play`?

**Comment by Sija** [comment]
@RX14 nope, that doesn't work.

**Comment by tymat** [comment]
> Well you can just
> 
> ```
> crystal play &
> nohup crystal play
> systemd-run --user crystal play
> ```

Not everyone uses systemd or Linux.

**Comment by tymat** [comment]
> I'm on the fence for user facing daemons.
> 
> For server daemons I fully agree with http://www.mikeperham.com/2014/09/22/dont-daemonize-your-daemons/
> 
> With systemd user sessions becoming a fairly common thing these days, I tend to think the same arguments start to apply here too.

This is not a good solution as it's not portable.  It works for Linux but not other platforms which do not use systemd to manage system processes.

**Comment by straight-shoota** [comment]
Systemd user services and the other mentioned methods are just examples to illustrate how you can easily achieve daemons with system tools. Not everything applies to all operating systems, but every os provides some tools for daemonizing processes. The point here is simply to recommend using such external tools, instead of implementing daemonizing directly in the playground. This task is simply much better handled by a system-specific external tool and doesn't need to be integrated into the playground.
